File: src/python/flows_specification/tasks.py
from __future__ import annotations
from harness_engine import docs_reader, harness_config, state_store
from . import evaluator, prompts, store

STEP_BUDGET = 18
SOURCES_FOLDER = prompts.SOURCES_FOLDER

# Phase -> prompt function, for resuming/advancing to a phase's own (non-retry) prompt.
_PHASE_PROMPTS = {
    "discover": prompts.discover_prompt,
    "product": prompts.product_prompt,
    "analysis": prompts.analysis_prompt,
    "design": prompts.design_prompt,
    "review": prompts.review_prompt,
    "approve": prompts.approve_prompt,
}


def start():
    r = store.load_run()
    resumable = (
        r.get("status") == "in_progress"
        and r.get("phase") in ("discover", "product", "analysis", "design", "review")
    ) or (r.get("status") == "awaiting_approval" and r.get("phase") == "approve")
    if resumable:
        return _PHASE_PROMPTS.get(r.get("phase"), prompts.discover_prompt)()

    store.reset()
    store.save_run(
        {
            "step": 0,
            "status": "in_progress",
            "phase": "discover",
            "counters": {},
            "traceLabel": None,
            "terminalReason": None,
        }
    )
    if docs_reader.has_docs(SOURCES_FOLDER):
        content, files = docs_reader.read(SOURCES_FOLDER)
        store.write_accepted("sources", {"files": files, "content": content})
    return prompts.discover_prompt()


def _advance(phase, prompt_fn):
    r = store.load_run()
    r.update({"step": state_store.load().step, "status": "in_progress", "phase": phase})
    store.save_run(r)
    return prompt_fn()


def discover(_=None):
    x = store.read_proposal("idea")
    if x is None:
        return prompts.discover_retry_prompt(
            [
                f"no readable idea proposal was found at '{prompts.IDEA_PROPOSAL_PATH}' "
                "(missing or not valid JSON)."
            ]
        )
    # Real provenance when `start` ingested a sources bundle; falls back to a fixed tag plus
    # a digest of the idea's own content when the run had no sources folder to draw from.
    sources, sources_digest = store.read_accepted("sources")
    files = (sources or {}).get("files") or []
    source = (
        f"{SOURCES_FOLDER} ({len(files)} file(s)): {', '.join(files)}"
        if files
        else "driver"
    )
    source_digest = sources_digest if sources is not None else store.digest(x)
    e = evaluator.idea(x, source, source_digest)
    if not e["passed"]:
        return prompts.discover_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    store.write_accepted("idea", x)
    return _advance("product", prompts.product_prompt)


def product(_=None):
    x = store.read_proposal("prd")
    if x is None:
        return prompts.product_retry_prompt(
            [
                f"no readable PRD proposal was found at '{prompts.PRD_PROPOSAL_PATH}' "
                "(missing or not valid JSON)."
            ]
        )
    _, parent = store.read_accepted("idea")
    e = evaluator.prd(x, parent or "")
    if not e["passed"]:
        return prompts.product_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    store.write_accepted("prd", x)
    return _advance("analysis", prompts.analysis_prompt)


def analysis(_=None):
    x = store.read_proposal("srs")
    if x is None:
        return prompts.analysis_retry_prompt(
            [
                f"no readable SRS proposal was found at '{prompts.SRS_PROPOSAL_PATH}' "
                "(missing or not valid JSON)."
            ]
        )
    p, pd = store.read_accepted("prd")
    e = evaluator.srs(x, pd or "", [g.get("id") for g in (p or {}).get("goals", [])])
    if not e["passed"]:
        return prompts.analysis_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    store.write_accepted("srs", x)
    return _advance("design", prompts.design_prompt)


def design(_=None):
    x = store.read_proposal("sdd")
    if x is None:
        return prompts.design_retry_prompt(
            [
                f"no readable SDD proposal was found at '{prompts.SDD_PROPOSAL_PATH}' "
                "(missing or not valid JSON)."
            ]
        )
    s, sd = store.read_accepted("srs")
    req = [
        r.get("id")
        for r in (s or {}).get("functionalRequirements", [])
        + (s or {}).get("qualityRequirements", [])
    ]
    e = evaluator.sdd(x, sd or "", req)
    if not e["passed"]:
        return prompts.design_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    store.write_accepted("sdd", x)
    return _advance("review", prompts.review_prompt)


def review(_=None):
    x = store.read_proposal("review")
    if x is None:
        return prompts.review_retry_prompt(
            [
                "no readable readiness verdict proposal was found at "
                f"'{prompts.REVIEW_PROPOSAL_PATH}' (missing or not valid JSON)."
            ]
        )
    s, _ = store.read_accepted("srs")
    d, _ = store.read_accepted("sdd")
    req = [
        r.get("id")
        for r in (s or {}).get("functionalRequirements", [])
        + (s or {}).get("qualityRequirements", [])
    ]
    adrs = [a.get("id") for a in (d or {}).get("adrs", [])]
    e = evaluator.readiness(x, req, adrs)
    if not e["passed"]:
        return prompts.review_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    r = store.load_run()
    if x.get("verdict") == "READY":
        store.write_accepted("readiness", x)
        r.update({"status": "awaiting_approval", "phase": "approve"})
        store.save_run(r)
        return "stop"
    count = r.setdefault("counters", {}).get("recascades", 0)
    if count >= 2:
        r.update(
            {
                "status": "needs_human_decision",
                "phase": "stop",
                "terminalReason": "recascade budget exhausted",
            }
        )
        store.save_run(r)
        return "stop"
    r["counters"]["recascades"] = count + 1
    target = x["verdict"].split(":", 1)[1]
    r.update({"status": "in_progress", "phase": target})
    store.save_run(r)
    prompt_fn = _PHASE_PROMPTS.get(target)
    if prompt_fn is None:
        return prompts.review_retry_prompt(
            [f"READINESS_VERDICT_INVALID: unroutable verdict '{x['verdict']}'"]
        )
    return prompt_fn()


def approve(_=None):
    x = store.read_proposal("approval")
    if x is None:
        return prompts.approve_retry_prompt(
            [
                "no readable approval proposal was found at "
                f"'{prompts.APPROVAL_PROPOSAL_PATH}' (missing or not valid JSON)."
            ]
        )
    current = store.bundle_digest()
    e = evaluator.approval(x, current)
    if not e["passed"]:
        return prompts.approve_retry_prompt(
            [f"{v['code']}: {v['message']}" for v in e["violations"]]
        )
    r = store.load_run()
    if x.get("decision") == "revise":
        r.update({"status": "in_progress", "phase": "review"})
        store.save_run(r)
        return prompts.review_prompt()
    prd, _ = store.read_accepted("prd")
    srs, _ = store.read_accepted("srs")
    sdd, _ = store.read_accepted("sdd")
    readiness, _ = store.read_accepted("readiness")
    if not all((prd, srs, sdd, readiness)):
        r.update(
            {
                "status": "publish_blocked",
                "phase": "stop",
                "terminalReason": "one or more accepted documents are missing",
            }
        )
        store.save_run(r)
        return "stop"

    from . import publisher

    rendered = publisher.render_all(prd, srs, sdd, readiness)
    gate = evaluator.development_readiness(
        prd,
        srs,
        sdd,
        readiness,
        x.get("bundleDigest", ""),
        current,
        rendered,
        harness_config.current().docs_max_chars,
    )
    if not gate["passed"]:
        reason = "; ".join(f"{v['code']}: {v['message']}" for v in gate["violations"])
        r.update(
            {"status": "publish_blocked", "phase": "stop", "terminalReason": reason}
        )
        store.save_run(r)
        return "stop"

    ok, error = publisher.publish(prd, srs, sdd, readiness)
    if not ok:
        r.update(
            {
                "status": "publish_blocked",
                "phase": "stop",
                "terminalReason": error or "publish postcondition failed",
            }
        )
        store.save_run(r)
        return "stop"
    r.update({"status": "completed", "phase": "stop"})
    store.save_run(r)
    return "stop"

File: src/rust/flows_specification/src/tasks.rs
//! Tasks (mirrors `SpecificationTasks.cs`): the state machine driving one phase transition
//! per call, wiring `store` (persistence), `evaluator` (validation), `renderer`/`publisher`
//! (publication) and `prompts` (driver-facing text) together.

use crate::evaluator::{validate_approval, validate_development_readiness, validate_idea, validate_prd, validate_readiness, validate_sdd, validate_srs};
use crate::prompts::{
    analysis_prompt, analysis_retry_prompt, approve_prompt, approve_retry_prompt, design_prompt,
    design_retry_prompt, discover_prompt, discover_retry_prompt, product_prompt,
    product_retry_prompt, review_prompt, review_retry_prompt,
};
use crate::publisher::publish;
use crate::renderer::render_all;
use crate::store::{
    DIR, SOURCES_FOLDER, accept, adr_ids_of, bundle_digest, digest, path, proposal, read,
    requirement_ids_of, run, save, save_status, write,
};
use harness_engine::Envelope;
use serde_json::json;
use std::fs;

pub(crate) const STEP_BUDGET: i32 = 18;

pub(crate) fn start() -> String {
    let r = run();
    let s = r["status"].as_str().unwrap_or("");
    let p = r["phase"].as_str().unwrap_or("");
    let resumable = (s == "in_progress"
        && ["discover", "product", "analysis", "design", "review"].contains(&p))
        || (s == "awaiting_approval" && p == "approve");
    if resumable {
        return match p {
            "product" => product_prompt(),
            "analysis" => analysis_prompt(),
            "design" => design_prompt(),
            "review" => review_prompt(),
            "approve" => approve_prompt(),
            _ => discover_prompt(),
        };
    }
    let _ = fs::remove_dir_all(DIR);
    write(
        "run.json",
        &json!({"step":0,"status":"in_progress","phase":"discover","counters":{}}),
    );
    if harness_engine::docs_reader::has_docs(SOURCES_FOLDER) {
        let (content, files) = harness_engine::docs_reader::read(SOURCES_FOLDER);
        accept("sources", &json!({"files": files, "content": content}));
    }
    discover_prompt()
}

pub(crate) fn discover(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("idea") else {
        return discover_retry_prompt(&[format!(
            "no readable idea proposal was found at '{}' (missing or not valid JSON).",
            path("idea.proposal.json")
        )]);
    };
    let sources = read("sources.accepted.json");
    let files: Vec<String> = sources
        .as_ref()
        .and_then(|s| s["files"].as_array())
        .into_iter()
        .flatten()
        .filter_map(|f| f.as_str().map(String::from))
        .collect();
    let source = if !files.is_empty() {
        format!("{SOURCES_FOLDER} ({} file(s)): {}", files.len(), files.join(", "))
    } else {
        "driver".to_string()
    };
    let source_digest = match &sources {
        Some(s) => digest(s),
        None => digest(&v),
    };
    let errors = validate_idea(&v, &source, &source_digest);
    if !errors.is_empty() {
        return discover_retry_prompt(&errors);
    }
    accept("idea", &v);
    save(run(), "product", "in_progress");
    product_prompt()
}

pub(crate) fn product(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("prd") else {
        return product_retry_prompt(&[format!(
            "no readable PRD proposal was found at '{}' (missing or not valid JSON).",
            path("prd.proposal.json")
        )]);
    };
    let Some(i) = read("idea.accepted.json") else {
        return product_retry_prompt(&["missing accepted idea.".to_string()]);
    };
    let errors = validate_prd(&v, &digest(&i));
    if !errors.is_empty() {
        return product_retry_prompt(&errors);
    }
    accept("prd", &v);
    save(run(), "analysis", "in_progress");
    analysis_prompt()
}

pub(crate) fn analysis(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("srs") else {
        return analysis_retry_prompt(&[format!(
            "no readable SRS proposal was found at '{}' (missing or not valid JSON).",
            path("srs.proposal.json")
        )]);
    };
    let Some(p) = read("prd.accepted.json") else {
        return analysis_retry_prompt(&["missing accepted PRD.".to_string()]);
    };
    let goals = p["goals"]
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(|goal| goal["id"].as_str().map(String::from))
        .collect::<Vec<_>>();
    let errors = validate_srs(&v, &digest(&p), &goals);
    if !errors.is_empty() {
        return analysis_retry_prompt(&errors);
    }
    accept("srs", &v);
    save(run(), "design", "in_progress");
    design_prompt()
}

pub(crate) fn design(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("sdd") else {
        return design_retry_prompt(&[format!(
            "no readable SDD proposal was found at '{}' (missing or not valid JSON).",
            path("sdd.proposal.json")
        )]);
    };
    let Some(s) = read("srs.accepted.json") else {
        return design_retry_prompt(&["missing accepted SRS.".to_string()]);
    };
    let requirements = requirement_ids_of(&s);
    let errors = validate_sdd(&v, &digest(&s), &requirements);
    if !errors.is_empty() {
        return design_retry_prompt(&errors);
    }
    accept("sdd", &v);
    save(run(), "review", "in_progress");
    review_prompt()
}

pub(crate) fn review(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("review") else {
        return review_retry_prompt(&[format!(
            "no readable readiness verdict proposal was found at '{}' (missing or not valid JSON).",
            path("review.proposal.json")
        )]);
    };
    let verdict = v["verdict"].as_str().unwrap_or("");
    let srs = read("srs.accepted.json").unwrap_or_default();
    let sdd = read("sdd.accepted.json").unwrap_or_default();
    let requirements = requirement_ids_of(&srs);
    let adrs = adr_ids_of(&sdd);
    let errors = validate_readiness(&v, &requirements, &adrs);
    if !errors.is_empty() {
        return review_retry_prompt(&errors);
    }
    if verdict == "READY" {
        accept("readiness", &v);
        save(run(), "approve", "awaiting_approval");
        return "stop".into();
    }
    let mut r = run();
    let current_count = r["counters"]["recascades"].as_i64().unwrap_or(0);
    if current_count >= 2 {
        save_status(r, "needs_human_decision", Some("recascade limit reached"));
        return "stop".into();
    }
    r["counters"]["recascades"] = json!(current_count + 1);
    let target = verdict.split(':').nth(1).unwrap_or("review");
    save(r, target, "in_progress");
    match target {
        "product" => product_prompt(),
        "analysis" => analysis_prompt(),
        "design" => design_prompt(),
        _ => review_retry_prompt(&[format!("READINESS_VERDICT_INVALID: unroutable verdict '{verdict}'")]),
    }
}

pub(crate) fn approve(_: Option<&Envelope>) -> String {
    let Some(v) = proposal("approval") else {
        return approve_retry_prompt(&[format!(
            "no readable approval proposal was found at '{}' (missing or not valid JSON).",
            path("approval.proposal.json")
        )]);
    };
    let current = bundle_digest();
    let errors = validate_approval(&v, &current);
    if !errors.is_empty() {
        return approve_retry_prompt(&errors);
    }

    let run_state = run();
    let decision = v["decision"].as_str().unwrap_or("");

    if decision == "revise" {
        save(run_state, "review", "in_progress");
        return review_prompt();
    }

    // decision == "approved" (validate_approval already restricted the allowed set).
    let prd = read("prd.accepted.json");
    let srs = read("srs.accepted.json");
    let sdd = read("sdd.accepted.json");
    let readiness = read("readiness.accepted.json");
    let (Some(prd), Some(srs), Some(sdd), Some(readiness)) = (prd, srs, sdd, readiness) else {
        save_status(
            run_state,
            "publish_blocked",
            Some("one or more accepted documents (prd/srs/sdd/readiness) are missing; cannot publish."),
        );
        return "stop".into();
    };

    // Pre-publish gate (blueprint 0006 "Antes de promover, o DevelopmentReadinessEvaluator
    // deve provar..."): rendered via the SAME render_all() a real publish() call uses, so this
    // byte-budget check sees exactly what would be written. specs/active/ is not touched if
    // this fails.
    let requirement_ids = requirement_ids_of(&srs);
    let adr_ids = adr_ids_of(&sdd);
    let rendered = render_all(&prd, &srs, &sdd, &readiness);
    let docs_max_chars = harness_engine::harness_config::current().docs_max_chars as i64;
    let bundle_digest_at_approval = v["bundleDigest"].as_str().unwrap_or("");
    let gate_errors = validate_development_readiness(
        &prd,
        &readiness,
        &requirement_ids,
        &adr_ids,
        bundle_digest_at_approval,
        &bundle_digest(),
        &rendered,
        docs_max_chars,
    );
    if !gate_errors.is_empty() {
        let reason = gate_errors.join("; ");
        save_status(run_state, "publish_blocked", Some(&reason));
        return "stop".into();
    }

    match publish(&prd, &srs, &sdd, &readiness) {
        Ok(_) => {
            save(run_state, "stop", "completed");
            "stop".into()
        }
        Err(reason) => {
            save_status(run_state, "publish_blocked", Some(&reason));
            "stop".into()
        }
    }
}
